### PR TITLE
WT-9201 change mixed-mode to a per-transaction basis

### PR DESCRIPTION
**Pull request by keithbostic**

### CODE REVIEW

**Comment by keithbostic** [comment]
@agorrod, @dgottlieb, this is the change to replaced the table leavel `mixed-mode` configuration with the per-transaction `no_timestamp` configuration.

There's still more work to be done, but I wanted to get your confirmation that this is what you want before finishing it up and opening up for more general review.

The specific changes that I'd like you to check are relatively small, just ccd1f15 and 1d650cd, those are the changes to the API and library code, you should be able to generally ignore the rest for now.

Thanks!

File: dist/api_data.py [hunk 1]
```diff
@@ -1735,6 +1731,10 @@ def __ge__(self, other):
         choices=['read-uncommitted', 'read-committed', 'snapshot']),
     Config('name', '', r'''
         name of the transaction for tracing and debugging'''),
+    Config('no_timestamp', 'false', r'''
```
Comment (dgottlieb) [inline dist/api_data.py:1734]: I think this is true of all txn flags, but just double checking: this flag is cleared when a transaction is committed/rolled back? And when the parent session is reset?

Comment (keithbostic) [reply dist/api_data.py:1734]: Yes; it's reinitialized for each new transaction.

**Review by dgottlieb — APPROVED** [review]

File: src/docs/arch-hs.dox [hunk 2]
```diff
@@ -85,7 +85,8 @@ isolation level and doesn't hold a valid snapshot.
 @section arch_hs_reconciliation History store and reconciliation
 When a dirty page is reconciled on a user file btree, the update chain is examined and the latest
 committed update is chosen as the on-disk value. All older updates are added to the history store
-table assuming they are not yet obsolete. Additionally any mixed mode timestamps will be applied.
+table assuming they are not yet obsolete. Additionally any updates without timestamps will be
+applied.
```
Comment (agorrod) [inline src/docs/arch-hs.dox:90]: IDK if it's auto-magic, but line wrapping looks wrong.

**Review by agorrod — APPROVED** [review]
I've got one tiny nit, otherwise LGTM.

I checked the API and name changes. I'm trusting that you got the logic right - I don't think I can check that via a review.

File: src/docs/arch-hs.dox (hunk 2 above)
Comment (keithbostic) [reply src/docs/arch-hs.dox:90]: Sorry, I'm missing something -- what looks wrong?

File: src/reconcile/rec_col.c [hunk 3]
```diff
@@ -1419,19 +1419,19 @@ __wt_rec_col_var(
...
                     break;
                 case WT_UPDATE_TOMBSTONE:
                     /*
-                     * When a mixed mode tombstone is getting written to disk, remove any historical
-                     * versions that are greater in the history store for this key.
+                     * When a tombstone without a timestamp is written to disk, remove any
+                     * historical versions that are greater in the history store for this key.
                      */
-                    if (upd_select.mm_tombstone && r->hs_clear_on_tombstone)
+                    if (upd_select.no_ts_tombstone && r->hs_clear_on_tombstone)
```
Comment (bvpvamsikrishna) [inline src/reconcile/rec_col.c:1434]: This is not a comment for this PR, but I think there might be an issue here in our current code.
@quchenhao , @kommiharibabu , it seems we are truncating the entries from the history store, if this a non timestamped tombstone with out considering the transaction id. But ideally we should truncate the history store entries only if this tombstone is globally visibile with respect to both timestamp and transaction id.

Comment (quchenhao) [reply src/reconcile/rec_col.c:1434]: Yes, I agree that is wrong.

Comment (kommiharibabu) [reply src/reconcile/rec_col.c:1434]: There is no wrong here. We return the tombstone from rec_upd_select only when the tombstone is globally visible. The flag `no_ts_tombstone` is just an indication of whether the tombstone is a no timestamped or not. Based on that we need to truncate the history store. Let me know If I miss anything.

Comment (bvpvamsikrishna) [reply src/reconcile/rec_col.c:1434]: Thanks for checking, I have scheduled an online review tmr to check this issue as the there are two different answers.
@keithbostic, sorry to hijack this PR for a day. As stated earlier, the issue in discussion is not from this PR. Please ignore my comments and proceed with this PR in your usual way. I will track the issue separately.

Comment (bvpvamsikrishna) [reply src/reconcile/rec_col.c:1434]: Just to close the loop, we ( myself, Hari & Chenhao) did an online review of the code, and are now satisfied that there is no issue with the logic.

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: src/docs/arch-hs.dox (hunk 2 above)
Comment (agorrod) [reply src/docs/arch-hs.dox:90]: Oh, sorry. It was my browser wrapping lines weirdly :(

File: src/history/hs_rec.c [hunk 4]
```diff
@@ -852,24 +849,24 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
     WT_ERR(ret);
 
     /*
-     * Fail the eviction if we detect mixed mode timestamps when we've passed the error return flag.
-     * We cannot modify the history store to fix the mixed mode timestamp updates as it may make the
-     * history store checkpoint inconsistent.
+     * Fail the eviction if we detect any timestamp ordering issue and the error flag is set. We
+     * cannot modify the history store to fix the updates' timestamps as it may make the history
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/history/hs_rec.c:853]: Thanks -- fixed in 458b07c62.

File: dist/api_data.py [hunk 5]
```diff
@@ -71,16 +70,13 @@ def __ge__(self, other):
...
-        are first used, and \c never which enforces that timestamps are never used for a table.
-        (The \c always, \c key_consistent and \c ordered choices should not be used, and are
-        retained for backward compatibility.)''',
-        choices=['always', 'key_consistent', 'mixed_mode', 'never', 'none', 'ordered']),
+        describe how timestamps are expected to be used on table modifications.  The choices are
+        the default, which ensures that once timestamps are used for a key, they are always used,
+        and also that multiple updates to a key never use decreasing timestamps and \c never which
+        enforces that timestamps are never used for a table.  (The \c always, \c key_consistent,
+        \c mixed_mode and \c ordered choices should not be used, and are retained for backward
+        compatibility.)''', choices=['always', 'key_consistent', 'mixed_mode', 'never', 'none',
```
Comment (quchenhao) [inline dist/api_data.py:78]: The mixed mode option here is for backward compatibility?

**Review by quchenhao — APPROVED** [review]
lgtm with a question

File: dist/api_data.py (hunk 5 above)
Comment (keithbostic) [reply dist/api_data.py:78]: Yes, MDB is still calling into the WT build with `mixed_mode` set.

**Pull request merged**